xclim/indices/run_length.py

```diff
@@ -56,13 +58,63 @@ def use_ufunc(
         If ufunc_1dim is "auto", returns True if the array is on dask or too large.
         Otherwise, returns ufunc_1dim.
     """
+    if ufunc_1dim is True and freq is not None:
+        raise ValueError(
+            "Resampling after run length operations is not implemented for 1d method"
+        )
+
     if ufunc_1dim == "from_context":
         ufunc_1dim = OPTIONS[RUN_LENGTH_UFUNC]
 
     if ufunc_1dim == "auto":
         ufunc_1dim = not uses_dask(da) and (da.size // da[dim].size) < npts_opt
+    # If resampling after run length is set up for the computation, the 1d method is not implemented
+    # Unless ufunc_1dim is specifically set to False (in which case we flag an error above),
+    # we simply forbid this possibility.
+    return (index == "first") and (ufunc_1dim) and (freq is None)
 
-    return index == "first" and ufunc_1dim
+
+def resample_and_rl(
+    da: xr.DataArray,
+    resample_before_rl: bool,
+    compute,
+    *args,
+    dim: str = "time",
+    freq: str | None = None,
+    **kwargs,
+) -> xr.DataArray | xr.Dataset:
+    """Wrap run length algorithms to control if resampling occurs before or after the algorithms.
+
+    If resample_before_rl is 'from_context', the parameter is read from xclim's global (or context) options.
+
+    Parameters
+    ----------
+    da: xr.DataArray
+      N-dimensional array (boolean).
+    resample_before_rl : bool
+      Determines whether if input arrays of runs `da` should be separated in period before
+      or after the run length algorithms are applied
+    compute
+      Run length function to apply
+    args
+      Positional arguments needed in `compute`
+    dim: str
+      The dimension along which to find runs.
+    freq : str
+      Resampling frequency.
+    kwargs
+      Keyword arguments needed in `compute`
+
+    Returns
+    -------
+    xr.DataArray
+      Output of compute resampled according to frequency {freq}.
+    """
+    if resample_before_rl and freq is not None:
+        out = da.resample({dim: freq}).map(compute, args=args, dim=dim, **kwargs)
+    else:
```

Comment:
```suggestion
    if resample_before_rl:
        out = da.resample({dim: freq}).map(compute, args=args, dim=dim, **kwargs)
    else:
```
If someone forgets `freq` but does give `resample_before_rl=True,` I believe it should fail. Here, the error will come from xarray's resample, we could also raise an explicit error by testing `if freq is None` _within_ the `if`.  The current code silently switches to resampling after if one forgets `freq`.

Reply:
I think I would agree if `resample_before_rl = True` was not the default option. I think the user should be able to specify  `freq is None` without having to care about `resample_before_rl`, because with `freq is None`,  resampling does not occur.

Reply:
Should we add an option `None` for resample_before_rl` and check that the user makes a coherent choice?

|`resample_before_rl` | `freq` | Result? |
|--|--|--|
| `None` | `None` | Success |
| `None` | "YS", "MS",... | Fail |
| `True`, `False` | `None` | Fail |
| `True`, `False` | "YS", "MS",... | Success |

Reply:
I didn't think of the case where one would want NOT to resample. However, wouldn't it be better for such a user to call the  `compute` function directly instead? I'm not sure I see a case where someone would use `resample_before_rl` in a way that needs to be "generic" to both resampling and non-resampling.

If such a case does exist, your idea is the best way to go!

Reply:
True, I'm not sure if that functionality would be used. 

Say you have a dataset over 30 years. Using `heat_wave_max_length(da,...,freq=None,...)` would allow you to find the maximal heat wave of your whole dataset directly. You could also use `heat_wave_max_length(da,...,freq="YS",..., resample_before_rl=False)` and  simply find the maximum in the set of 30 `heat_wave_max_length` sub-output for each year

Reply:
On a side note, I realized that the signature in the indices is:
``freq: str = "YS"` 
as opposed to letting  `None`  as a possibility
``freq: str | None = "YS"` .

I will wait to see what we decide to do, but if we keep `freq=None`, I'll need to fix this.